Spring framework and Spring boot post

mckornfield opened the pull request:
* Added a post about Spring boot and the Spring
framework, which hits some important high level points
* Added a couple images and a slight css tweak to center
one of them

Commit:
Spring framework and Spring boot post

* Added a post about Spring boot and the Spring
framework, which hits some important high level points
* Added a couple images and a slight css tweak to center
one of them

mckornfield commented:
@michaelkochub Ready for review. If you are wondering what the image looks like: 
![image](https://user-images.githubusercontent.com/16839335/51771761-116df180-20b8-11e9-8a0c-5e48303e6159.png)

mckornfield commented:
I think it is slightly offcenter, but center-align did nothing X/

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+Hi again from guest writer Matt here! When I say Spring to you, and you think of the season (i.e. rabbits and Easter and such) first and not the Java technology, then this post will probably be very enlightening for you! If you know what Spring is, then hopefully you will learn something.
+
+The tl;dr for the Spring Framework is that it is a massive set of Java libraries. We are going to focus only on a couple pieces of Spring for this piece, Spring IoC (Inversion of Control) and Spring Boot.
```
nit - colon instead of comma
```markdown
We are going to focus only on a couple pieces of Spring for this piece: Spring IoC (Inversion of Control) and Spring Boot.
```

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+Before we talk about Spring, a bit about the terminology.
+
+## What is IoC (Inverison of Control)?
```
`inversion`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+**Dependency Injection** Dependencies are provided at runtime, not compile time, by an application container. Ask and you shall receive.
+
+**Dependency Inversion** A code style, where abstractions do not depend on details, and objects only depend on objects at their same level of abstraction or higher. Keep things on a need to know basis and use interfaces/ implementations to separate layers of abstraction.
```
`interfaces/implementations`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+**Dependency Inversion** A code style, where abstractions do not depend on details, and objects only depend on objects at their same level of abstraction or higher. Keep things on a need to know basis and use interfaces/ implementations to separate layers of abstraction.
+
+What do these look like you may ask? Let's try and explain without using programming, and instead use a dog walker, let's call him David.
```
`walker; let's call him David`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+What do these look like you may ask? Let's try and explain without using programming, and instead use a dog walker, let's call him David.
+
+David can walk any species of Dog, as long as the dogs are provided to David. That means someone outside of David is responsible for delivering the dogs to him to walk. Let's call that "entity" that is responsible for delivering the dogs, Spring.
```
no need for `,`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+David can walk any species of Dog, as long as the dogs are provided to David. That means someone outside of David is responsible for delivering the dogs to him to walk. Let's call that "entity" that is responsible for delivering the dogs, Spring.
+
+So Spring is responsible for waking David up in the morning, handing him the dogs, and saying "Okay David, now you can walk the dogs." David then walks the dogs, in a way Spring is completely unaware of. Once David is done, he returns the Dogs to Spring and waits until the next time Spring asks him to walk the dogs. Maybe David has a side hustle as an uber driver, but Spring doesn't care.
```
lowercase `dogs`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+In this way, David had to wait until he was handed the dogs, so we can say he had a _dependency_ on the dogs, which was _injected_ i.e. provided by an outside context that David knew very little about.
+
+Lastly, Spring did not care about anything David did with the dogs. Spring didn't have to depend on _how_ David walked the dogs, just that David was a Dog Walker and he needed Dogs to walk. In fact, maybe David is sick one day, so Spring instead brings the dogs to Frank, another dog walker. Spring, nor any other part of the situation (the dogs, the people who own the dogs, etc.) are dependent on  ...
```
variable capitalization... I see `Dog Walker` and `dog walker` and `Dog` and `dog`. Is there a special meaning behind this?

mckornfield replied:
No... I think I was just confusing classes and the anecdote. I will remove this Unnecessary Capitalization.

michaelkochub replied:
Thank you Very much.

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+In this way, David had to wait until he was handed the dogs, so we can say he had a _dependency_ on the dogs, which was _injected_ i.e. provided by an outside context that David knew very little about.
+
+Lastly, Spring did not care about anything David did with the dogs. Spring didn't have to depend on _how_ David walked the dogs, just that David was a Dog Walker and he needed Dogs to walk. In fact, maybe David is sick one day, so Spring instead brings the dogs to Frank, another dog walker. Spring, nor any other part of the situation (the dogs, the people who own the dogs, etc.) are dependent on  ...
```
situation? Maybe clarify `Spring, nor any other party that plays a role in this situation`.

Also I sense a run-on. I think `walking. The only requirement here is that Frank and David fulfill their contract as dog walkers.` would be a little better

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+## What does this look like in Spring framework code?
+
+There are essentially two ways to create an application context, one way is via Bean Creation + Importing, the other via Injection/ Auitowiring of dependencies. I'll explain each in some detail.
```
`Bean Creation and Importing`, `/Autowiring`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+### Explicit declaration of beans (@Bean)
+
+The Spring framework, for many versions of Java, dependend heavily on XML, primarly because Java did not have any way to pepper metadata into classes. Once annotations were introduced in Java 5, there was a bit of a move away from XML defined beans and into direct annotations. For an example of XML based bean declaration, see [this medium post][medium post on spring xml]
```
`depended`

michaelkochub replied:
`primarily`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+The Spring framework, for many versions of Java, dependend heavily on XML, primarly because Java did not have any way to pepper metadata into classes. Once annotations were introduced in Java 5, there was a bit of a move away from XML defined beans and into direct annotations. For an example of XML based bean declaration, see [this medium post][medium post on spring xml]
+
+Continuing on with our example, let's say I want to make this wonderful app, here are some code snippets of how I might start out, using the annotation approach instead of the XML or autoconfiguration approaches:
```
`app. Here`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
````diff
@@ -0,0 +1,336 @@
+Continuing on with our example, let's say I want to make this wonderful app, here are some code snippets of how I might start out, using the annotation approach instead of the XML or autoconfiguration approaches:
+
+```java
+  public interface DogWalker{}
````
but would this pass checkstyle 😉

mckornfield replied:
in my repo there ain't no checkstyle 🔫 🔫 pew pew

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
````diff
@@ -0,0 +1,336 @@
+  public interface DogWalker{}
+```
+```java
+  public interface Dog{}
````
no space after the class name... you're turning into sam 

`Dog {}`

mckornfield replied:
The master has become the student

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
````diff
@@ -0,0 +1,336 @@
+```
+
+```java
+  public class David implements DogWalker{
````
same for all these `DogWalker {`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
````diff
@@ -0,0 +1,336 @@
+  }
+```
+
+So this is our wonderful app, which doesn't really do anything except set up a skeleton for us, we are ready to build our Spring context. All we need to do is set up a class to house our beans in, annotated with a `@Configuration`
````
```
With our wonderful app, which doesn't really do anything except set up a skeleton for us, we are ready to build our Spring context.
```

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+Spring will be much happier in this case, and will be able to look up the dog by name from its registry.
+
+The last piece of this toy problem that is worth considering is, what if we wanted to add more dogs? It might start to get messy if we have one file that just has 50 different dogs in it. Spring allows us to separate out our configs, and as long as we include them in our application at runtime, we can declare them in any number of Config files. We can also import them into our current config, so  ...
```
`The last piece of this toy problem that is worth considering is what if we wanted to add more dogs?`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
````diff
@@ -0,0 +1,336 @@
+}
+```
+
+That about wraps up our discussion of explicit dependencies. How might you use this in a java app? You would need to either have this Config imported via an xml (for example, with Tomcat) or an annotation, but this leads more into Spring Boot, which we'll cover in a bit.
````
`Java`, or maybe capitalization doesn't matter

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+## Some final black magic
+
+So spring can manage all your dependency with the drop of an annotation. But what about really large applications? What about testing? Spring is very well set up to support pulling in beans to tests, using things like `@SpringBootTest`. But there are plenty of [tutorials][spring boot testing] that you can use to flex your testing muscles. Testing with Spring is worth its own post.
```
`So spring can manage all your dependencies with the annotations as your workhorses.` Or something, idk. pluralize `dependency` at the very least 😏

mckornfield replied:
lol snarky grammarian

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+For `@PostConstruct`, you can annotate a method of a class declared as a bean, and this method will be called after calling the constructor and supplying any autowired dependencies.
+
+For `@PreDestroy`, this will be called before the bean is wiped out (either during shutdown or if the bean is scoped differently). Usually better for Beans that hold onto a resource, or maybe have a different scope besides Singleton.
```
`This is usually better...`

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -0,0 +1,336 @@
+
+A sample application that has a bit more than the code snippets above can be [found here][sample mck repo].
+
+Spring is big and if you take the time to figure out what's going on, you can make it work for you.
```
powerful conclusion 👏

michaelkochub commented on _sass/main.scss:
```diff
@@ -384,3 +384,6 @@ section {
   }
 }
+.align-center{
+  margin-left: 30% // This is hacky
```
Matt... _blog poster-cum-css pro_

mckornfield replied:
Lol is this is latin for plz remove this, or are you giving me an accolade?

michaelkochub replied:
accolade

michaelkochub reviewed (commented):
some typos here and there but overall solid... I'll be quicker with the CR next time

mckornfield commented:
@michaelkochub Addressed feedback, good old vim does not have the best spellcheck though I turned it on. I just inlined some html for the image so I didn't have to do that hacky css thingy

michaelkochub commented on _posts/2019-1-24-spring-is-black-magic.md:
```diff
@@ -302,13 +303,13 @@ There is a lot going on in there few lines. The most important thing to understa
 ## Some final black magic
 
-So spring can manage all your dependency with the drop of an annotation. But what about really large applications? What about testing? Spring is very well set up to support pulling in beans to tests, using things like `@SpringBootTest`. But there are plenty of [tutorials][spring boot testing] that you can use to flex your testing muscles. Testing with Spring is worth its own post.
+So spring can manage all your dependencies with annotations as you workhorses. But what about really large applications? What about testing? Spring is very well set up to support pulling in beans to tests, using things like `@SpringBootTest`. But there are plenty of [tutorials][spring boot testing] that you can use to flex your testing muscles. Testing with Spring is worth its own post.
```
are you calling the reader a workhorse? One 'r' can do a lot.

`annotations as your workhorses`

mckornfield replied:
damn spellcheck doesn't know how bad my grammar is

michaelkochub reviewed (approved):
one final typo, but overall looks good. feel free to squash and I will merge.
